lib: add toposort, nixos: use toposort for fileSystems to properly support bind and move mounts

oxij opened the pull request.

oxij committed:
lib: introduce listDfs and toposort, add example to hasPrefix

oxij committed:
nixos: apply toposort to fileSystems to support bind and move mounts

And use new `config.system.build.fileSystems` property everywhere.

oxij committed:
nixos: tasks/fileSystems: cleanup

garbas commented on lib/lists.nix:
```diff
@@ -193,6 +193,38 @@ rec {
+  # Depth-First Search (DFS) for `list != []`.
+  #
+  # `before a b == true` means that `b` depends on `a` (there's an
+  # edge from `b` to `a`).
```
could you provide an example of usage for listDfs

garbas commented on lib/lists.nix:
```diff
@@ -193,6 +193,38 @@ rec {
+  #
+  # `before a b == true` means that `b` should be after `a`
+  # in the result.
+  toposort = before: list:
```
could you provide an example of usage for toposort

oxij replied:
You mean in the comment there? Because the next patch in the list is a
good example in itself.

garbas replied:
yes, in comment. majority of this functions have nice examples which help a lot when going through them.

oxij replied:
> yes, in comment. majority of this functions have nice examples which help a lot when going through them.

Done. Majority doesn't, btw.

oxij commented:
I hope it's okay now.

lethalman commented:
Cool, an automatic dependency discovery for mountpoints. So how do you solve this:

```
mount somedev /foo
mount /foo/image /bar
mount --bind /bar /foo
```

Actual dependencies:

```
/foo -> /bar -> /foo/image -> /foo -> foodev
```

There's a cycle there, I don't think your toposort can solve it.

The problem is: mounting is a problem that cannot be solved simply by automatic dependency discovery. It's just as not expressive as the mounting problem.

oxij commented:
I don't. My primary use case is mount binding simple sequences in initrd
(complicated netboot).

But notice that your example also can't be expressed with fileSystems:

    fileSystems = {
      "/" = { fsType = "tmpfs"; };
      "/foo" = { device = "/dev/some";
                 fsType = "xfs"; };
      "/bar" = { device = "/foo/image";
                 fsType = "xfs"; };
      "/foo" = { device = "/bar";
                 fstype = "none";
                 options = "bind";
               };
    };

repeats "/foo".

I can, however, imagine a way to implement this by a kind of explicit
"version" field in `fileSystems.*` submodule, or, more NixOS-sy,
describing actual mountpoint explicitly and making newer "/foo" receive
higher "version" than old "/foo", e.g.:

    fileSystems = {
      "/" = { fsType = "tmpfs"; };
      "/foo-v1" = { mountpoint = "/foo";
                    device = "/dev/some";
                    fsType = "xfs"; };
      "/bar"    = { device = "/foo/image";
                    fsType = "xfs"; };
      "/foo-v2" = { mountpoint = "/foo";
                    needs  = [ "/foo-v1" ];
                    device = "/bar";
                    fstype = "none";
                    options = "bind";
                  };
    };

What you've reminded me is that I didn't think about other types of
cycles which might hang `nix-instantiate` with current implementation. I
need to check this.

lethalman commented:
It's just easier to define after and before systemd options in the mount options, instead of trying to be clever.

oxij commented:
1) In initrd?

2) So what? Both ways you'll have to give the same information to either
   this toposort or systemd which will do its own toposort inside. Which
   means changing `fileSystems` in some way or another.

I won't oppose if someone adds proper `after` and `before`s to systemd
units in NixOS, it just won't solve my problem in initrd, so I did this
(and applied the same thing to /etc/fstab to solve it for free in the
rest of the system, but I can remove that line if you think that no
solution at all is better).

lethalman commented:
1) It's still a broken implementation. Systemd should go in initrd. In the meanwhile you can as well mount stuff yourself with initrd hooks.

2) No your toposort is an improvement but it's not a solution, and might as well lead to other problems in ordering.

oxij commented:
(Edit's don't get mailed, can you please not use them except for typos,
not whole paragraphs?)

My point is that this PR is not broken, because

1) it solves a problem that can not be solved in any other way except
   putting the whole systemd into `initrd` (I would be happy if someone
   does this, I waited long enough, but, AFAIK, that effort is dead),

2) any other solution is bound to be as "broken" since it's gonna use
   the same inputs and algorithms.

oxij commented:
> Clever discovery of dependencies in this context just doesn't work.

Yes, that's why I don't do it.

Let's be constructive. Show me an example that current `fileSystems` can
express and `toposort`ing it breaks something.

oxij commented:
> What you've reminded me is that I didn't think about other types of
> cycles which might hang `nix-instantiate` with current implementation.
> I need to check this.

Okay, I had been thinking better than I though I was. `listDfs` ignores
cycles, so current implementation `toposort`s cycles as if they start
from arbitrary node.

AFAIK, it doesn't break anything that is not already broken in user's
`configuration.nix`, but I can make it into an evaluation error. Should
I?

garbas commented:
@oxij tnx for the examples, that make it easier to understand.

i'd wait for @lethalman opinion wheather to raise an evaluation error.

copumpkin commented:
I'm in favor of a partial solution that can accurately detect when it doesn't know what to do and throw an evaluation error. Silently doing the wrong thing (i.e., what we have today) always makes me uncomfortable.

lethalman commented:
I'm still not in favor of this sorry. I'd rather put some manual ordering in `fileSystems`, before and after, which could apply both to initrd fstab and systemd mounts as well.

jagajaga commented:
I agree with @oxij and @copumpkin. This solution saves our "lives" sometimes. Now we don't have _any_ solution at all. So it's better to have this than to have nothing.

rasendubi commented:
(triage) Any update on systemd in initrd? As far as I understand, we still don't have it.

I think we should decide if this PR is ok to merge (after rebasing) or should be rejected.

abbradar commented:
I want to have `toposort` in lib to use set instead of a list in #17822 and to nicely integrate hidepid there. Also I don't think that it would hurt having it for `fileSystems`, too -- it won't make things worse and may help somebody. At least it's some determined and logical order that we can document and expect people to understand its limitations.

@oxij, do you want to rebase it by yourself? I'm willing to do the job if you lack motivation;
@lethalman, are you still opposed to this? My point for having it is that it allows to cleanup things more (see #17822) and gives some benefits -- I don't expect systemd in initrd on any short term (I actually have tried to add it and dracut, but got enormous initrd size increase and lots of broken FHS assumptions and dropped the idea).

oxij commented:
Thanks for showing some interest! I thought this was dead.

Problems with the implementation presented here:

* it doesn't generate errors on encountered cycles (I have a local
  version that does that),

* `fsBefore` is, strictly speaking, incorrect (algebraically, UNIX paths
  are lists separated via "/", not strings, e.g. "/dev/ab" is not a
  parent of "/dev/abc"). I'm pretty sure I can fix that.

* Pretty much no interest till now.

Anyway, do you need anything else for sets in #17822 (which is pretty
nice, btw) or the current state rebased onto master suffices?

abbradar commented:
Problems you list can't really do anything worse that it is now, and it covers my usecase still, so you can just rebase for now. Thanks!

edolstra commented:
Well, this function uses `++` so it probably has O(n^2) performance. IMHO this kind of processing is better done in the builder or at runtime.

edolstra commented:
It's probably possible to do this efficiently using `builtins.genericClosure`. Still, it's better to deal with ordering at runtime (as systemd already does - this is only an issue because we don't have systemd in the initrd).

abbradar commented:
I agree that doing this in systemd sounds nicer (and would abolish need in #17822, too, because mount options would be used right from systemd), but I don't think we will have it anytime soon ~_^.

Doing this in builder may be an option if we want to reduce computations in Nix, but I think it's a good idea to try and optimize this first (we already have the code, after all). I'm not sure how can `genericClosure` help but it looks interesting! Unfortunately it's [not documented](https://github.com/NixOS/nix/issues/552) but the [source code](https://github.com/NixOS/nix/blob/master/src/libexpr/primops.cc#L284) helps.

oxij commented:
> Well, this function uses `++` so it probably has O(n^2) performance.

In the worst case, assuming I (++) is lazy.

> It's probably possible to do this efficiently using
> `builtins.genericClosure`.

I'll look into that.

In any case, I didn't notice the evaluation time of this implementation
in my local branch for n=4..10. Not to mention the fact that NixOS
already has `strings-with-deps.nix` which does some crazy (++) stuff (I
can't really understand what it does exactly) that gets evaluated all
the time too.

> Still, it's better to deal with ordering at runtime (as systemd
> already does - this is only an issue because we don't have systemd in
> the initrd).

Personally, I fail to see why one would want to push for
systemd-specific NixOS instead of making NixOS more generic when
possible. E.g. lately I dream of having an an option to use something
simpler in place of systemd, e.g. OpenRC, because every second systemd
update breaks something on at least one of my machines, and policykit
and journald drive my crazy. (And in an ideal world, I'd want to have an
initd that evaluates nix-expressions.)

oxij committed:
nixos: copy resize2fs only for stage-1 fileSystems

oxij commented:
Fun fact while rebasing: 3d16af70bf894ce15ec9bdcad3c9ac736dc43630 sorts
fileSystems for completely irrelevant reasons so that "/" gets to be the
first one.

I say this shows that they should be topologically sorted in
nix-expressions anyway.

oxij committed:
Revert a soon to be useless pice of "nixos/stage-1: add mechanism which lustrates all impurities from / (#17784)"

This reverts a pice of commit 3d16af70bf894ce15ec9bdcad3c9ac736dc43630.

oxij committed:
nixos: generalize copy-paste from stage-1 and zfs to utils

oxij commented:
Pushed the reworked branch.

Now it generates pretty assertions, e.g.:

~~~~
{lib, pkgs, config, ...}: with lib;
{

    boot.loader.grub.device = "/dev/sda";

    fileSystems = {
      "/" = { fsType = "tmpfs"; };
      "/foo" = { device = "/bar";
                 fsType = "none";
                 options = "bind";
               };
      "/bar" = { device = "/foo";
                 fsType = "none";
                 options = "bind";
               };
    };

}
~~~~

produces:

~~~~
Failed assertions:
- The ‘fileSystems’ option can't be topologically sorted: mountpoint dependency path /bar -> /foo loops to /bar
~~~~

Also, I think it should actually speed evaluation up a bit on average,
since I fixed all the things to toposort and extract fileSystems attrset
to list only once, instead of like ten times.

Also cleaned up some copy-paste.

Evaluates ok here.

I hope I didn't break #17784, would be nice if @obadz could check that.

abbradar commented:
I've revieved an updated patchset and it looks okay to me. I haven't run any tests yet.

@edolstra, do you think we need to optimize this further before a merge?

I'm going to merge in few days if everyone is okay with it, but before I'll conduct several tests with updated #17822 on top of that to ensure things are okay. Also would be nice to get more reviews -- I haven't got a close look at `toposort` and `listDfs`, assuming their correctness for now (I'll do that later).

obadz commented on nixos/modules/system/boot/stage-1.nix:
```diff
@@ -128,21 +134,6 @@ let
-      (fs: fs.mountPoint != "/" && (fs.neededForBoot || elem fs.mountPoint [ "/nix" "/nix/store" "/var" "/var/log" "/var/lib" "/etc" ]));
-    filterRoot = filter
-      (fs: fs.mountPoint == "/");
-    allFileSystems = attrValues config.fileSystems;
```
Pls make sure that `/` remains the first fs to be mounted by default, otherwise this will break #17784.
Users who have fancy setups and decide to change that ordering give up the ability to lustrate…

obadz commented:
> Fun fact while rebasing: 3d16af7 sorts fileSystems for completely irrelevant reasons so that "/" gets to be the first one.

What makes you think the reasons are irrelevant?

abbradar commented:
@obadz I think what @oxij means is that we already did ad-hoc toposort for #17784 (which is irrelevant to #17822 that also needs it).

oxij commented:
AFAI can prove, toposorting should push `/` to the top of the list in
all the situations I can think of, since everything should be mounted
somewhere under `/` and `fsBefore` compares on `mountPoint`s.

Yep, I meant "irrelevant [to the current discussion] reasons" and didn't
even think that out of context that could be read as derogatory. Sorry
about that.

abbradar commented:
Tested with my own system, verified that `/` is first and that ordering is correct and also did another review.

@oxij Thanks for your work!

oxij commented:
You're welcome. Thanks for making the sane approach prevail :)
